core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -388,6 +393,32 @@ public void updatePrimaryTerm(final long newPrimaryTerm) {
                         e -> failShard("exception during primary term transition", e));
                 primaryTerm = newPrimaryTerm;
                 latch.countDown();
+                // prevent primary relocation handoff while resync is not completed
+                primaryReplicaResyncInProgress = true;
```

Comment:
I tried it first using a permit. The issue is that there is no guarantee that the permit is taken before the permit acquired by the relocation handoff.
If I try to take the permit at the current line, then it will be added to the delayedOperations. Once fillSeqNoGaps completes and releases the delayed operations, they will try to reacquire the permit on a generic thread, which will be racing against the relocation handoff.

I'm open for suggestions here.